# Introducao e Exmplos

In [15]:
test = [5, 1, 2, 3, 4]

print(max(test))
print(min(test))
print(sum(test))

5
1
15


In [3]:
print(isinstance("Python", (int,str)))
print(isinstance(10.5,int))

True
False


In [16]:
print(list(filter(lambda x: x > 2, test)))

print(list(map(lambda x: x * 2, test)))

[5, 3, 4]
[10, 2, 4, 6, 8]


In [22]:
print(list(reversed(test)))

print(list(sorted(test)))

print(list(sorted(test, reverse = 1)))

[4, 3, 2, 1, 5]
[1, 2, 3, 4, 5]
[5, 4, 3, 2, 1]


## Funcao recursiva

São aquelas funções que chamam a si mesmas para resolver problemas
repetitivos até atingir uma condição de parada

In [23]:
def factorial(n):
    if n == 0:
        return 1
    return n * factorial(n - 1)

print(factorial(5))

120


### GPT explanation of the execution of the function above

First call: factorial(5) is called. Since n is not 0, it moves to return 5 * factorial(4).

But before it can multiply, it needs to know what factorial(4) is, so it calls factorial(4).

Second call: factorial(4) is now called. Again, n isn't 0, so it returns 4 * factorial(3).

Again, it needs to know what factorial(3) is, so it calls factorial(3).

Third call: factorial(3) is called. It returns 3 * factorial(2), so it calls factorial(2).

Fourth call: factorial(2) is called. It returns 2 * factorial(1), so it calls factorial(1).

Fifth call: factorial(1) is called. It returns 1 * factorial(0), so it calls factorial(0).

Sixth call: factorial(0) is called. Now, the base case is hit: it immediately returns 1, because if n == 0: return 1.

---

Now that factorial(0) returns 1, the chain of unfinished multiplications starts resolving:

factorial(1) now becomes 1 * 1 = 1.

factorial(2) now becomes 2 * 1 = 2.

factorial(3) now becomes 3 * 2 = 6.

factorial(4) now becomes 4 * 6 = 24.

factorial(5) now becomes 5 * 24 = 120.

Finally, print(factorial(5)) prints 120.

---

To answer your question:
Yes, once n reaches 0 and returns 1, it "unwinds" through all the previous incomplete multiplications, one by one, finishing the chain backwards.

## Lambda

In [24]:
soma = lambda a, b: a + b

print(soma(3,5))

8


## Funcoes com closures

Closure é uma função interna que lembra do ambiente onde foi criada, mesmo
após a execução da função externa ter terminado. Isso permite que ela mantenha
o estado de variáveis sem a necessidade de usá-las globalmente.

In [25]:
def mulitplicador(n):
    def multiplicado(x):
        return x * n
    return multiplicado

triplo = mulitplicador(3)
valor = triplo(5)
print(valor)

15


In [27]:
def age():
    b_year = int(input("Insert the birth year: "))
    c_year = int(input("Insert the current year: "))
    age = c_year - b_year
    print(f"They are {age} years old.")
    
age()

They are 27 years old.


2nd Idea

In [ ]:
def age(year1, year2):
    return abs(year1 - year2)

input_01, input_02 = (input("Insert the current year and the year of birth: ").split())

print(age(int(input_01),int(input_02)))

27


In [29]:
def letter_count(word):
    return len(word)

w_input = input("Insert the word you want to know the amount of letters in: ")
length = letter_count(w_input)

print(f"Your word has {length} letters")

Your word has 6 letters


2nd idea, without using len()

In [ ]:
def word_length(word):
    n = 0
    b = True

    while b:
        try:
            b = bool(word[n])
        except:
            b = False
            break
        n += 1
    return n

input_w = word_length(input("Insert your word: "))

print("Your word has", input_w,"words")

Your word has 7 words


In [51]:
def greetings(time):
    if time < 12:
        return "Bom dia!"
    elif 12 <= time < 18:
        return "Boa Tarde!"
    else:
        return "Boa Noite!"
    
time = int(input("Insert what time it is (24h format): "))

print(greetings(time))

Boa Noite!


In [ ]:
telefones = ["11987654321", "21912345678", "31987654321", "11911223344"] 

def convert_str2int(num_list):
    list(map(lambda num: num = int(num), num_list))
    return num_list

print(convert_str2int(telefones))        

['11987654321', '21912345678', '31987654321', '11911223344']
